avr-hal-generic/src/spi.rs

```diff
@@ -0,0 +1,202 @@
+                    // set up data order control bit
+                    match self.settings.data_order {
+                        DataOrder::MostSignificantFirst => w.dord().clear_bit(),
+                        DataOrder::LeastSignificantFirst => w.dord().set_bit(),
+                    };
+                    // set up polarity control bit
+                    match self.settings.mode.polarity {
+                        spi::Polarity::IdleHigh => w.cpol().set_bit(),
+                        spi::Polarity::IdleLow => w.cpol().clear_bit(),
+                    };
+                    // set up phase control bit
+                    match self.settings.mode.phase {
+                        spi::Phase::CaptureOnFirstTransition => w.cpha().clear_bit(),
+                        spi::Phase::CaptureOnSecondTransition => w.cpha().set_bit(),
+                    };
+                    // set up clock rate control bit
+                    match self.settings.clock {
+                        SerialClockRate::OscfOver2 => w.spr().val_0x00(),
+                        SerialClockRate::OscfOver4 => w.spr().val_0x00(),
+                        SerialClockRate::OscfOver8 => w.spr().val_0x01(),
+                        SerialClockRate::OscfOver16 => w.spr().val_0x01(),
+                        SerialClockRate::OscfOver32 => w.spr().val_0x02(),
+                        SerialClockRate::OscfOver64 => w.spr().val_0x02(),
+                        SerialClockRate::OscfOver128 => w.spr().val_0x03(),
+                    }
+                });
+                // set up 2x clock rate status bit
+                self.peripheral.spsr.write(|w| match self.settings.clock {
+                    SerialClockRate::OscfOver2 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver4 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver8 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver16 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver32 => w.spi2x().set_bit(),
+                    SerialClockRate::OscfOver64 => w.spi2x().clear_bit(),
+                    SerialClockRate::OscfOver128 => w.spi2x().set_bit(),
+                });
+            }
+        }
+
+        /// FullDuplex trait implementation, allowing this struct to be provided to
+        /// drivers that require it for operation.  Only 8-bit word size is supported
+        /// for now.
+        impl $crate::hal::spi::FullDuplex<u8> for $Spi {
+            type Error = $crate::void::Void;
+
+            /// Sets up the device for transmission and sends the data
+            fn send(&mut self, byte: u8) -> $crate::nb::Result<(), Self::Error> {
+                if !self.is_write_in_progress {
+                    self.is_write_in_progress = true;
+                    self.write(byte);
+                }
+                self.assert_write_complete()?;
+                self.is_write_in_progress = false;
+                Ok(())
```

Comment:
I think you haven't quite hit the nail in how `nb` works.  When `send()` returns `WouldBlock` it means the requested operation cannot be completed right now and should be retried at a later time.  This means, the _same_ byte which was just attempted should be retried.

What your code does right now is this:  You push the byte to the hardware and then return `WouldBlock`, indicating to the caller that it should retry sending that exact same byte.  It will then retry until `send()` returns `Ok(())` and only then start with the next byte.

While I think your code will at least not send any wrong data out to the bus, it does violate the `nb` contract a bit.  A caller would think the byte was not sent (queued in the hardware) yet which means it will be busy retrying a number of times even though it was already sent (queued).  Ideally, the retrying only happens for the follow-up byte, which can't be sent (queued) until the previous one was flushed out.

To do this, you want to first check whether the previous write (if any) completed and only then start sending the follow-up:

```suggestion
                if !self.is_write_in_progress {
                    self.flush()?;
                    self.is_write_in_progress = false;
                }
                self.write(byte);
                self.is_write_in_progress = true;
                Ok(())
```

Reply:
Interesting, okay.  It sounds like nb is NOT (as I thought) a framework for ensuring that a task is completed.  Rather, it is a framework for ensuring a task is not executed until the system is ready to handle it.

If I apply your suggestion, I will also need to alter `read()` to check is_write_in_progress too.  I'll work on this.